server/engine/engine.py

```diff
@@ -1125,6 +1217,182 @@ def processLocalSubjectFolder(path: str, outputName: str = None, href: str = '')
+                        f.write('\n')
+                        for frame in badDynamicsFrames['torqueDiscrepancy']:
+                            f.write(f'     - frame {frame} (time = {timestamps[frame]:.3f})\n')
+                        reasonNumber += 1
+
+                    if 'notOverForcePlate' in badDynamicsFrames:
+                        f.write('\n')
+                        f.write(f'{reasonNumber}. "Foot not over a force plate."\n')
+                        f.write(f'   ------------------------------\n')
+                        f.write(textwrap.indent(textwrap.fill(
+                            'External ground reaction forces were present and the foot was detected '
+                            'to be penetrating the ground, but the foot was not located over any '
+                            'known force plate in the following frames:'), '   '))
+                        f.write('\n')
+                        for frame in badDynamicsFrames['notOverForcePlate']:
+                            f.write(f'     - frame {frame} (time = {timestamps[frame]:.3f})\n')
+                        reasonNumber += 1
+
+                    if 'missingImpact' in badDynamicsFrames:
+                        f.write('\n')
+                        f.write(f'{reasonNumber}. "Missing foot-ground impact detected."\n')
+                        f.write(f'   --------------------------------------\n')
+                        f.write(textwrap.indent(textwrap.fill(
+                            'EXPERIMENTAL: The following frames had ground reaction force data removed using an impact '
+                            'detection algorithm in nimblephysics. If you receive this message, then we assume that '
+                            'you know what you are doing.'), '   '))
+                        f.write('\n')
+                        for frame in badDynamicsFrames['missingImpact']:
+                            f.write(f'     - frame {frame} (time = {timestamps[frame]:.3f})\n')
+                        reasonNumber += 1
+
+                    if 'missingBlip' in badDynamicsFrames:
+                        f.write('\n')
+                        f.write(f'{reasonNumber}. "Missing ground reaction force \'blip\' detected."\n')
+                        f.write(f'   ------------------------------------------------\n')
+                        f.write(textwrap.indent(textwrap.fill(
+                            'Ground reaction forces were detected the following frames were preceded and followed by '
+                            'several frames of zero force data. Therefore, these \'blips\' in the data were removed:')
+                            , '   '))
+                        f.write('\n')
+                        for frame in badDynamicsFrames['missingBlip']:
+                            f.write(f'     - frame {frame} (time = {timestamps[frame]:.3f})\n')
```

Comment:
This list especially could get quite long and repetitive, since the frames will be consecutive by construction. Perhaps we should present all the missing GRF frames lists as ranges, rather than lists?

Reply:
Done!